src/ads/vast/client/VASTClient.js

```diff
@@ -32,174 +38,116 @@ VASTClient.prototype.getVASTResponse = function getVASTResponse(adTagUrl, callba
       cb(e);
     }
   }
-};
 
-VASTClient.prototype._sendVASTResponse = function sendVASTResponse(callback) {
-  var that = this;
-  callback = callback || noop;
+  function sanityCheck(adTagUrl, cb) {
+    if (!adTagUrl) {
+      return new VASTError('on VASTClient.getVASTResponse, missing ad tag URL');
+    }
 
-  return function (error, response) {
-    if (error) {
-      vastUtil.track(that.errorURLMacros, {ERRORCODE: error.code || 900});  //900 <== Undefined error
+    if (!isFunction(cb)) {
+      return new VASTError('on VASTClient.getVASTResponse, missing callback function');
     }
-    callback(error, response);
-  };
+  }
 };
 
-VASTClient.prototype._getAd = function getVASTAd(adTagUrl, callback) {
-  var error;
+VASTClient.prototype._getVASTAd = function (adTagUrl, callback) {
   var that = this;
-  var options = isObject(adTagUrl) && !isNull(adTagUrl) ? adTagUrl : {adTagUrl: adTagUrl};
-  options.ads = options.ads || [];
-  error = sanityCheck(options, callback);
-  if (error) {
-    if (isFunction(callback)) {
-      return callback(error, null);
-    }
-    throw error;
-  }
 
-  async.waterfall([
-    requestVASTXml,
-    buildAd
-  ], callback);
-
-  /*** local function ***/
-  function sanityCheck(opts, cb) {
-    if (!opts.adTagUrl) {
-      return new VASTError('on VASTClient._getAd, missing ad tag URL');
+  getAdWaterfall(adTagUrl, function (error, vastTree) {
+    var waterfallAds = vastTree && isArray(vastTree.ads) ? vastTree.ads : null;
+    if (error) {
+      that._trackError(error, waterfallAds);
+      return callback(error, waterfallAds);
     }
 
-    if (!isFunction(cb)) {
-      return new VASTError('on VASTClient._getAd, missing callback function');
-    }
+    getAd(waterfallAds.shift(), [], waterfallHandler);
 
-    if (opts.ads.length >= that.WRAPPER_LIMIT) {
-      return new VASTError("on VASTClient._getAd, players wrapper limit reached (the limit is " + that.WRAPPER_LIMIT + ")", 302);
+    /*** Local functions ***/
+    function waterfallHandler(error, adChain) {
+      if (error) {
+        that._trackError(error, adChain);
+        if (waterfallAds.length > 0) {
+          getAd(waterfallAds.shift(),[], waterfallHandler);
```

Comment:
after I would call return;

so you can remove the else and have only callback(error, adChain)